bsmetadata/metadata_utils.py

```diff
@@ -183,14 +183,33 @@ def _collate_metadata(metadata_list: List[dict], cfg: MetadataConfig):
         assert metadata_node.relative_start_pos not in metadata_dict_idx[metadata_node.char_start_idx]
         assert metadata_node.relative_end_pos not in metadata_dict_idx[metadata_node.char_end_idx]
 
-        metadata_dict_idx[metadata_node.char_start_idx][metadata_node.relative_start_pos] = start_text
-        metadata_dict_idx[metadata_node.char_end_idx][metadata_node.relative_end_pos] = end_text
+        if start_text:
+            metadata_dict_idx[metadata_node.char_start_idx][metadata_node.relative_start_pos] = start_text
+        if end_text:
+            metadata_dict_idx[metadata_node.char_end_idx][metadata_node.relative_end_pos] = end_text
 
     for absolute_idx, value in metadata_dict_idx.items():
         pos_sorted = sorted(list(value.keys()))
         local_metadata = ""
         for pos in pos_sorted:
             local_metadata += metadata_dict_idx[absolute_idx][pos]
+
+        # We add here a local special token if needed around the metadata list of a type if needed
+        if (
+            cfg.local_metadata_special_token_start
+            and local_metadata
+            and metadata_list[0]["key"] in cfg.local_metadata_special_token_start
+        ):
+            local_special_token_start = cfg.local_metadata_special_token_start[metadata_list[0]["key"]]
+            local_metadata = f"{local_special_token_start}{local_metadata}"
+        if (
+            cfg.local_metadata_special_token_end
+            and local_metadata
+            and metadata_list[0]["key"] in cfg.local_metadata_special_token_end
+        ):
+            local_special_token_end = cfg.local_metadata_special_token_end[metadata_list[0]["key"]]
+            local_metadata = f"{local_metadata}{local_special_token_end}"
+
```

Comment:
I'm not sure if we've discussed this earlier and I just forgot about it, but this seems to be very specific to entity metadata. Wouldn't it be better then to remove this logic from the main processing and instead add it to the custom metadata processor? Or is there some other kind of metadata that would also require this?

Reply:
I also take the opportunity to share my thoughts on this point when we talked about it with Manan (I remember asking the exact same questions initially, it's funny!). By the way, the only trace on github of our exchange with Manan at the time is [this PR](https://github.com/manandey/metadata/pull/1) . :slightly_smiling_face: 

In the end, I was thinking that this framework could allow to do what you had proposed to do a while ago : frame local tags with tag detectors to be able to remove them. 

Cf, the notes of the 2021 Sep 17 meeting with the following example: 
```
use special token: This is a [BEGIN_MD type=html] <b> [END_MD] great [BEGIN_MD type=html] </b> [END_MD] apple [BEGIN_MD type=entity] {{Apple (fruit)}} [END_MD]
```